### ISSUE TITLE
Trivy - The first run cannot skip downloading DB

### ISSUE THREAD

**Issue by lassehastrup** [oxsecurity/megalinter#4408]
**Describe the bug**
We're using oxsecurity/megalinter:latest in our devOps pipeline.
Previously we've had issues with trivy with the error described below, as it seems to be a general issue from their side regarding "downloading DB".

It's important to state, that it's far from every time we encounter this issue. This might only be on 5% of our pipeline runs.

From my understanding, your team have made changes to the way trivy was downloaded by caching trivy in the docker image, so that it didn't have to download from trivy on every run.

** Error received **

[Trivy] Hit TOOMANYREQUESTS: try again
[Trivy] Hit TOOMANYREQUESTS 5 times: Run trivy with --skip-db-update and --skip-check-update
❌ Linted [REPOSITORY] files with [trivy]: Found 1 error(s) - (22.18s)
- Using [trivy v0.57.1] https://megalinter.io/8.3.0/descriptors/repository_trivy
- MegaLinter key: [REPOSITORY_TRIVY]
- Rules config: identified by [trivy]
- Commands:
  [trivy fs --scanners vuln,misconfig --exit-code 1 .]
  [trivy fs --scanners vuln,misconfig --exit-code 1 .]
  [trivy fs --scanners vuln,misconfig --exit-code 1 .]
  [trivy fs --scanners vuln,misconfig --exit-code 1 .]
  [trivy fs --scanners vuln,misconfig --exit-code 1 .]
  [trivy fs --scanners vuln,misconfig --exit-code 1 .]
  [trivy fs --scanners vuln,misconfig --exit-code 1 . --skip-db-update --skip-check-update]
--Error detail:
2024-12-17T19:07:10Z	ERROR	[vulndb] The first run cannot skip downloading DB
2024-12-17T19:07:10Z	FATAL	Fatal error	init error: DB error: database error: --skip-update cannot be specified on the first run

docker pull oxsecurity/megalinter:latest | docker run

**Comment by nvuillam** [comment]
@lassehastrup damn, they were supposed to have fixed the problem with a new mirror gcr.io :/

https://github.com/aquasecurity/trivy/discussions/7668

I wonder if the fact that we try to call trivy unsuccessfully just before makes the workaround fail... :/

**Comment by ReneRebsdorf** [comment]
Any thoughts/updates on this @nvuillam ? :-)

I think you are right wrt. to preliminary calls, but I am not sure.

**Comment by nvuillam** [comment]
@ReneRebsdorf I did not investigate yet :/
For the moment, I just run again the jobs... but I agree that's a far from ideal situation

**Comment by nvuillam** [comment]
@ReneRebsdorf I also see it from time to time, but indeed that's not often

**Comment by ReneRebsdorf** [comment]
@nvuillam I would like to try to reproduce this, and see if it is indeed something that causes a fail before the cache is used, or what is going on.

As this will be the first time I look into the megalinter src, could you give some pointers? I am guessing it is related to the [trivy linter python file](https://github.com/oxsecurity/megalinter/blob/main/megalinter/linters/TrivyLinter.py) and the [trivy docker file](https://github.com/oxsecurity/megalinter/blob/main/linters/repository_trivy/Dockerfile), any other areas/files that should have special attention?

**Comment by nvuillam** [comment]
@ReneRebsdorf TrivyLinter class is indeed where it is handled :)

If you want to debug, you can locally install trivy on your computer then run trivy python test classes :)

**Comment by ReneRebsdorf** [comment]
@nvuillam just looked into this briefly.

As you probably already know this code is what we are seeing in this incident:

![Image](https://github.com/user-attachments/assets/69bc3979-cfaf-4dc6-a5ad-7adf9c1694e5)

I find it strange that the error we get from trivy is `The first run cannot skip downloading DB`, as the docker image should already have that as per running `trivy image --download-db-only --no-progress` in `linters/repository_trivy/Dockerfile`

Maybe the root cause relates to a wrongful/missing cache dir?
My only exposure to trivy is through megalinter where you guys have kindly taken care of all the intricacies :-)

Thoughts @nvuillam ?

**Comment by nvuillam** [comment]
@ReneRebsdorf your analysis is good, when we build the image we store the cache in the image (i don't know where but it shouldn't matter as we keep all the files)

I suspect a trivy bug, because as we have the cache we shouldn't see `The first run cannot skip downloading DB`

Also note that we try that only after 5 `too many requests` so the main issue remains that their database has still availability issues , otherwise it wouldn't go through the workaround implemented in MegaLinter

**Comment by ReneRebsdorf** [comment]
I agree but nonetheless there is a need for the workaround and thus it should work, if possible

I read on the trivy issue tracker there were problems previously with cache not being specified which caused problems, if that is the case here a fix for the workaround might be easy.

I will have a look, time permitting, but quite a bit hung up at the moment.

**Comment by ReneRebsdorf** [comment]
I looked into it at the time of writing but have been unable to reproducerede.

Let's close and we can reopen if need be?
